File: src/main/java/com/github/liebharc/cachetable/CacheTableIndex.java
package com.github.liebharc.cachetable;

import org.h2.command.dml.AllColumnsForPlan;
import org.h2.engine.Session;
import org.h2.index.*;
import org.h2.message.DbException;
import org.h2.result.Row;
import org.h2.result.RowImpl;
import org.h2.result.SearchRow;
import org.h2.result.SortOrder;
import org.h2.table.Column;
import org.h2.table.IndexColumn;
import org.h2.table.TableFilter;
import org.h2.value.*;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CacheTableIndex extends BaseIndex {

    /**
     * The index of the indexed column.
     */
    private final Column[] indexColumn;
    private final CombinedCacheMetaData metaInfo;
    private final static ArrayList<Value[]> emptyResult = new ArrayList<>();

    public CacheTableIndex(
            CacheTable table,
            int id,
            String indexName,
            IndexColumn[] columns,
            IndexType indexType,
            CombinedCacheMetaData metaInfo) {
        super(table, id, indexName, columns, indexType);
        this.metaInfo = metaInfo;
        indexColumn = new Column[metaInfo.getNumberOfIndexColumns()];
        for (int i = 0; i < indexColumn.length; i++) {
            indexColumn[i] = columns[i].column;
        }
    }

    @Override
    public void truncate(Session session) {
        // Changes aren't supported
    }

    @Override
    public void add(Session session, Row row) {
        // Changes aren't supported
    }

    @Override
    public void remove(Session session, Row row) {
        // Changes aren't supported
    }

    @Override
    public Cursor find(Session session, SearchRow first, SearchRow last) {
        if (first == null || last == null) {
            return findAll(session);
        }

        Value[] firstValue = getValue(first);
        Value[] lastValue = getValue(last);

        if (Arrays.equals(firstValue, lastValue)) {
            return findSingleValue(session, firstValue);
        }

        return findRangeValue(session, firstValue, lastValue);
    }

    private Value[] getValue(SearchRow row) {
        Value[] result = new Value[indexColumn.length];
        for (int i = 0; i < result.length; i++) {
            Column col = indexColumn[i];
            Value value = row.getValue(col.getColumnId());
            result[i] = value != null ? value.convertTo(col.getType(), database.getMode(), null) : null;
        }

        return result;
    }

    @NotNull
    private Cursor findSingleValue(Session session, Value[] firstValue) {
        /*
         * Sometimes the incoming search is a similar, but not the same type
         * e.g. the search value is INT, but the index column is LONG. In which
         * case we need to convert, otherwise the HashMap will not find the
         * result.
         */
        List<Value[]> values = metaInfo.getRowOrNull(firstValue, session);
        return createCursor(values);
    }

    private Cursor createCursor(List<Value[]> values) {
        if (values == null) {
            return new CacheCursor(emptyResult.iterator());
        }

        if (values.size() == 1) {
            return new SingleRowCursor(new RowImpl(values.get(0), 0));
        }

        return new CacheCursor(values.iterator());
    }

    private Cursor findRangeValue(Session session, Value[] firstValue, Value[] lastValue) {
        return new CacheCursor(metaInfo.getRowsInRange(session, firstValue, lastValue).iterator());
    }

    private Cursor findAll(Session session) {
        return new CacheCursor(metaInfo.getAllRows(session));
    }

    @Override
    public long getRowCount(Session session) {
        return getRowCountApproximation();
    }

    @Override
    public long getRowCountApproximation() {
        return metaInfo.getRowSize();
    }

    @Override
    public long getDiskSpaceUsed() {
        return 0;
    }

    @Override
    public void close(Session session) {
        // nothing to do
    }

    @Override
    public void remove(Session session) {
        // nothing to do
    }

    @Override
    public double getCost(Session session, int[] masks,
                          TableFilter[] filters, int filter, SortOrder sortOrder,
                          AllColumnsForPlan allColumnsSet) {
        if (masks == null) {
            return 2;
        }

        for (Column column : columns) {
            int index = column.getColumnId();
            int mask = masks[index];
            if ((mask & IndexCondition.EQUALITY) != IndexCondition.EQUALITY) {
                return Long.MAX_VALUE;
            }
        }
        return 2;
    }

    @Override
    public void checkRename() {
        // ok
    }

    @Override
    public boolean needRebuild() {
        return false;
    }

    @Override
    public boolean canGetFirstOrLast() {
        return false;
    }

    @Override
    public Cursor findFirstOrLast(Session session, boolean first) {
        throw DbException.getUnsupportedException("HASH");
    }

    @Override
    public boolean canScan() {
        return true;
    }
}

File: src/main/java/com/github/liebharc/cachetable/CacheTableIndex.kt
package com.github.liebharc.cachetable

import com.google.common.cache.Cache
import org.h2.command.dml.AllColumnsForPlan
import org.h2.engine.Database
import org.h2.engine.DbObject
import org.h2.engine.Session
import org.h2.index.*
import org.h2.result.Row
import org.h2.result.RowImpl
import org.h2.result.SearchRow
import org.h2.result.SortOrder
import org.h2.schema.Schema
import org.h2.table.Column
import org.h2.table.IndexColumn
import org.h2.table.Table
import org.h2.table.TableFilter
import org.h2.value.ValueLong
import org.h2.value.ValueString
import java.lang.StringBuilder
import java.util.ArrayList

class CacheTableIndex(val cache: Cache<Long, String>, val table: CacheTable) : Index {
    override fun getComment(): String {
        return "Cache Index"
    }

    override fun getId(): Int {
        return 0
    }

    override fun isFindUsingFullTableScan(): Boolean {
        return false
    }

    override fun getColumns(): Array<Column> {
        return columns
    }

    override fun getPlanSQL(): String {
        return "Cache index"
    }

    override fun truncate(session: Session?) {
    }

    override fun getRowCountApproximation(): Long {
        return cache.size()
    }

    override fun getCreateSQLForCopy(table: Table?, quotedName: String?): String {
        return "Cache index"
    }

    override fun findNext(session: Session?, higherThan: SearchRow?, last: SearchRow?): Cursor {
        TODO("not implemented") //To change
    }

    override fun rename(newName: String?) {
    }

    override fun compareRows(rowData: SearchRow?, compare: SearchRow?): Int {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun close(session: Session?) {
    }

    override fun setSortedInsertMode(sortedInsertMode: Boolean) {
    }

    override fun getRowCount(session: Session?): Long {
        return cache.size()
    }

    override fun checkRename() {
    }

    override fun createLookupBatch(filters: Array<out TableFilter>?, filter: Int): IndexLookupBatch {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun setComment(comment: String?) {
    }

    override fun find(session: Session?, first: SearchRow?, last: SearchRow?): Cursor {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun find(filter: TableFilter?, first: SearchRow?, last: SearchRow?): Cursor {
        if (first != null && last != null && first.getValue(0).`object` == last.getValue(0).`object` ) {
            val key = last.getValue(0).`object` as Int
            val longKey = key.toLong()
            return SingleRowCursor(RowImpl(arrayOf(ValueLong.get(longKey), ValueString.get(cache.getIfPresent(longKey)!!)), 0))
        }

        return CacheCursor(cache.asMap().entries.iterator())
    }

    override fun isTemporary(): Boolean {
        return false
    }

    override fun isRowIdIndex(): Boolean {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getDiskSpaceUsed(): Long {
        return cache.size()
    }

    override fun getName(): String {
        return "CacheIndex"
    }

    override fun canGetFirstOrLast(): Boolean {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getDatabase(): Database {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun canFindNext(): Boolean {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun update(session: Session?, oldRow: Row?, newRow: Row?) {
        // Cache updates are not supported
    }

    override fun add(session: Session?, row: Row?) {
        // Cache updates are not supported
    }

    override fun getCreateSQL(): String {
        return "Column index"
    }

    override fun getRow(session: Session?, key: Long): Row {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getCost(session: Session?, masks: IntArray?, filters: Array<out TableFilter>?, filter: Int, sortOrder: SortOrder?, allColumnsSet: AllColumnsForPlan?): Double {
        return 1.0
    }

    override fun isHidden(): Boolean {
        return false
    }

    override fun getSchema(): Schema {
        return table.schema
    }

    override fun getChildren(): ArrayList<DbObject> {
        return ArrayList()
    }

    override fun getDropSQL(): String {
        return "Drop caches is not allowed"
    }

    override fun isFirstColumn(column: Column?): Boolean {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun getTable(): Table {
        return table
    }

    override fun getType(): Int {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun remove(session: Session?, row: Row?) {
    }

    override fun remove(session: Session?) {
    }

    override fun getSQL(alwaysQuote: Boolean): String {
        return "Column index"
    }

    override fun getSQL(builder: StringBuilder?, alwaysQuote: Boolean): StringBuilder {
        return builder!!
    }

    override fun needRebuild(): Boolean {
        return false
    }

    override fun removeChildrenAndResources(session: Session?) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun findFirstOrLast(session: Session?, first: Boolean): Cursor {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun setTemporary(temporary: Boolean) {
    }

    override fun canScan(): Boolean {
        return false
    }

    override fun getColumnIndex(col: Column?): Int {
        return table.columns.indexOf(col)
    }

    override fun getIndexColumns(): Array<IndexColumn> {
        return IndexColumn.wrap(table.columns)
    }

    override fun getIndexType(): IndexType {
        return IndexType.createUnique(true, true)
    }

}